### PR TITLE
feat: Calculate token usage statistics using call stack #117

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 1]
```diff
@@ -129,11 +147,72 @@ public String toString() {
...
+        public synchronized void link(String parentSpanId, String childSpanId) {
+            CallInfo callInfo = spanIdToCallInfo.get(parentSpanId);
+            if (callInfo == null) {
+                log.warn("Parent span is not found by id: {}", parentSpanId);
+                return;
+            }
+            callInfo.childSpanIds.add(childSpanId);
+        }
+
+        public synchronized void setTokeUsage(String spanId, TokenUsage tokenUsage) {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java:179]: Please correct type: setToke**n**Usage

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 2]
```diff
@@ -81,22 +85,36 @@ public RateLimitResult limit(ProxyContext context) {
...
         }
-        return entity != null;
+        entity.addCall(context.getCurrentSpanId());
+        return result;
+    }
+
+    public void calculateTokenUsage(ProxyContext context) {
+        Entity entity = getEntityFromTracingContext(context);
+        if (entity == null) {
+            log.warn("Entity is not found by traceId={}", context.getTraceId());
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java:114]: Please add prefix to locate the problem more easily: "Failed to calculate token usage"

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 3]
```diff
@@ -129,11 +147,72 @@ public String toString() {
...
+        }
+
+        public synchronized void addCall(String spanId) {
+            spanIdToCallInfo.put(spanId, new CallInfo());
+        }
+
+        public synchronized void link(String parentSpanId, String childSpanId) {
+            CallInfo callInfo = spanIdToCallInfo.get(parentSpanId);
+            if (callInfo == null) {
+                log.warn("Parent span is not found by id: {}", parentSpanId);
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java:173]: Please improve logging in this class.

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 4]
```diff
@@ -81,22 +85,36 @@ public RateLimitResult limit(ProxyContext context) {
...
+        return result;
+    }
+
+    public void calculateTokenUsage(ProxyContext context) {
+        Entity entity = getEntityFromTracingContext(context);
+        if (entity == null) {
+            log.warn("Entity is not found by traceId={}", context.getTraceId());
+            return;
+        }
+        context.setTokenUsage(entity.calculate(context.getCurrentSpanId()));
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java:117]: Is it better to do it only if it is missing? Some applications/models can compute it more precisely.

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java [hunk 5]
```diff
@@ -129,11 +147,72 @@ public String toString() {
...
 
-    private record Entity(String id, List<String> roles, String name, boolean user) {
+    @Data
+    private static class Entity {
+
+        private final String id;
+        private final List<String> roles;
+        private final String name;
+        private final boolean user;
+        private final Map<String, CallInfo> spanIdToCallInfo = new HashMap<>();
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/limiter/RateLimiter.java:157]: Please move the logic related to token tracking to a separate class. e.g. TokenTracker

File: src/main/java/com/epam/aidial/core/limiter/RateLimiter.java (hunk 4 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/limiter/RateLimiter.java:117]: We can rely on applications/assistants any longer to compute statistics

**Comment by astsiapanay** [comment]
no longer valid since the core is moved to per request key authorization

**Pull request closed without merging**